File: src/model/fourier_pe.rs
use std::f64::consts::TAU;

use tch::{
    Tensor,
    nn::{self, Module},
};

#[derive(Debug)]
// following https://arxiv.org/abs/2103.15813 (PixelTransformer) and https://arxiv.org/abs/2006.10739 (Fourier Features Positional Encoding)
pub struct PositionEncoder {
    projection: Tensor,
}

impl PositionEncoder {
    pub fn new(path: nn::Path, in_dim: i64, out_dim: i64) -> Self {
        Self {
            projection: path
                .randn("proj", &[in_dim, out_dim / 2], 0., 2.)
                .requires_grad_(false),
        }
    }
}

impl Module for PositionEncoder {
    fn forward(&self, pos: &Tensor) -> Tensor {
        let stub = (pos * TAU).matmul(&self.projection);
        Tensor::cat(&[stub.cos(), stub.sin()], -1)
    }
}

File: speckle-py/speckle/model/fourier_pe.py
import math

import torch
from torch import nn

TAU = 2.0 * math.pi


class PositionEncoder(nn.Module):
    def __init__(self, in_dim: int, out_dim: int):
        super().__init__()
        proj = torch.empty(in_dim, out_dim // 2).normal_(0.0, 2.0)
        self.register_buffer("proj", proj)

    def forward(self, pos: torch.Tensor) -> torch.Tensor:
        stub = (pos * TAU) @ self.proj
        return torch.cat([stub.cos(), stub.sin()], -1)
